### PR TITLE
Spawn cron jobs for match threads creation

### PR DESCRIPTION
**Pull request by dilico**
I've started working on implementing the bot as a REST API that can be invoked by cron jobs.

The idea is that we have one cron job ('Schedule') that runs every day at 00:05 UTC time (if I'm not wrong the football-data API returns kick off times in UTC). If there's a match on that day, the job creates a new cron job ('Pre-Match') that runs 6 hours before kick off.

The 'Pre-Match' cron job creates the Pre-Match thread on Reddit, deletes itself and creates another cron job ('Match') that runs 30 minutes before kick off.

The 'Match' cron job creates the Match thread on Reddit, deletes itself and creates another cron job ('Live-Match') that runs every minute.

The 'Live-Match' cron job follows the match minute by minute, so it can be used to notify of changes (e.g. score and cards) and to create a Post-Match thread on Reddit when the match is over (btw could you please add a method on the scraping that can be used to check if the match is finished?).

The PR is mainly to give an idea of what I'm talking about, but it would be great to hear your opinion about the general approach.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,16 +24,15 @@
...
     "@types/node-schedule": "^2.1.0",
     "jest": "^29.6.2",
     "ts-jest": "^29.0.0-next.1",
     "ts-node": "^10.9.1",
+    "ts-node-dev": "^2.0.0",
     "tsconfig-paths": "^4.2.0",
     "typescript": "^5.1.6"
-  },
-  "_moduleAliases": {
-    "@milanbot": ".build/src"
```
Comment (Kaizor) [inline package.json:35]: Can we keep this, please? We'll need it when it is built.

File: package.json [hunk 2]
```diff
@@ -2,18 +2,20 @@
   "name": "milanbot",
   "version": "1.0.0",
   "description": "",
-  "main": "index.js",
+  "main": "dist/index.js",
   "scripts": {
-    "build": "tsc --project tsconfig.json --outDir .build",
+    "build": "tsc",
     "test": "jest",
     "start": "npx ts-node -r tsconfig-paths/register src/index.ts",
```
Comment (Kaizor) [inline package.json]: As we have dev now, it'd be good to update start to be the thing that starts the built version :) Happy to do this in another PR tho

File: src/routing/index.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  res.json({
+    message: error.message
+  });
+}) as ErrorRequestHandler;
+
+export default {
+  errorHandler,
+  scheduleRequestHandler,
+  preMatchRequestHandler,
+};
```
Comment (Kaizor) [inline src/routing/index.ts]: Common practice here for an index file like this would be to export them individually instead of as default. e.g.
```
export { scheduleRequestHandler } from '@milanbot/routing/schedule-handler';
export { preMatchRequestHandler } from '@milanbot/routing/pre-match-handler';

export const errorHandler = ...
```

File: tsconfig.json [hunk 4]
```diff
@@ -29,7 +29,7 @@
 
     /* Modules */
     "module": "commonjs" /* Specify what module code is generated. */,
-    "rootDir": "./" /* Specify the root folder within your source files. */,
+    "rootDir": "./src" /* Specify the root folder within your source files. */,
```
Comment (Kaizor) [inline tsconfig.json:32]: rootDir needs to be './' so we can access the package.json to get the package version. This is used when doing API calls to reddit.

File: src/index.ts [hunk 5]
```diff
@@ -1,47 +1,15 @@
-import dotenv from "dotenv";
-import "module-alias/register";
-import schedule from "node-schedule";
-import FixturesFactory from "./services/fixtures/factory";
-import PreMatch from "./services/thread/pre-match";
+import dotenv from 'dotenv';
+import express from 'express';
+import routing from '@milanbot/routing';
```
Comment (Kaizor) [inline src/index.ts]:
```suggestion
import { scheduleRequestHandler, preMatchRequestHandler, errorHandler } from '@milanbot/routing';
```
When importing named exports, it is best to export them like this and be explicit about it.

File: src/routing/pre-match-handler.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    const fixture = new Fixture(fixtureType);
+    await createPreMatchThread();
+    await scheduler.deletePreMatchCronJob();
+    await scheduler.createMatchCronJob(fixture, `${req.protocol}://${req.get('host')}`);
+    res.status(201);
+    res.send();
+  };
+  handlePreMatchRequest()
+    .catch(next);
+}) as RequestHandler;
```
Comment (Kaizor) [inline src/routing/pre-match-handler.ts]: For all the handlers, we shouldn't use the promises like this. We should use try/catch.

```suggestion
export const preMatchRequestHandler = async (req: Request, res: Response, next: NextFunction) => {
  console.log("Received pre-match request");

  try {  
    const fixtureType = req.body as FixtureType;
    const fixture = new Fixture(fixtureType);
    
    await createPreMatchThread();    
    await scheduler.deletePreMatchCronJob();    
    await scheduler.createMatchCronJob(fixture, `${req.protocol}://${req.get('host')}`);
    
    res.sendStatus(201);
  } catch (error: unknown) {
     console.error(error);
     next();
  }
};
```

File: src/routing/schedule-handler.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  const fixtureService = new FixturesFactory().create();
+  return await fixtureService.getFixtures(new Date());
+};
+
+export const scheduleRequestHandler = (async (req, res, next) => {
+  console.log("Received schedule request");
+  const handleScheduleRequest = async () => {
+    const fixtures = await getTodaysFixtures();
+    if (!fixtures.length) {
+      res.status(204);
```
Comment (Kaizor) [inline src/routing/schedule-handler.ts:15]: This wouldn't send the response back, lets replace it with `sendStatus`

File: src/routing/schedule-handler.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      return;
+    }
+    if (fixtures.length > 1) {
+      res.status(400);
```
Comment (Kaizor) [inline src/routing/schedule-handler.ts:19]: Same as above, `sendStatus`

File: src/services/fixtures/fixture.ts [hunk 9]
```diff
@@ -15,7 +15,7 @@ class Fixture {
     return this.fixture.away;
   }
 
-  public get kickOff() {
+  public get kickOff(): String {
```
Comment (Kaizor) [inline src/services/fixtures/fixture.ts]:
```suggestion
  public get kickOff(): string {
```

File: src/scheduler/index.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+const cronJobUrl = 'https://api.cron-job.org';
+
+const createPreMatchCronJob = async (fixture: Fixture, url: string) => {
+  const config = {
+    headers: {
+      'Authorization': `Bearer ${process.env.CRON_JOB_API_KEY}`,
+      'Content-Type': 'application/json',
+    },
+  };
```
Comment (Kaizor) [inline src/scheduler/index.ts]: I see this is repeated in all the functions, we could pull it out and define it once in the module scope above the functions

File: src/scheduler/index.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+      },
+      'requestMethod': 1,
+      'notification': {
+        'onFailure': true,
+      },
+      'extendedData': {
+        'body': fixture.stringify(),
+      },
+    },
+  };
```
Comment (Kaizor) [inline src/scheduler/index.ts]: Can we make an unexported private helper function to create this payload, passing the function the title, url, kickoff, and fixture? As we use it in the MatchCronJob as well.

File: src/scheduler/index.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  if (response.status != 200) {
+    throw new Error(`Failed creating match cron job [${response.status}]`);
+  }
+};
+
+export default {
+  createPreMatchCronJob,
+  deletePreMatchCronJob,
+  createMatchCronJob,
+};
```
Comment (Kaizor) [inline src/scheduler/index.ts]: Same as before, normally just export the functions instead of doing the export default. Just allows for functions to move around more easily and doesn't couple the module

File: package.json (hunk 1 above)
Comment (dilico) [reply package.json:35]: Absolutely! Forgot to put it back when I was messing around with the build (see also below). [cyclic.sh](https://cyclic.sh) might need `index.js` at the root (and not in a subfolder), and I was trying to avoid leaking package.json too. I've now created `version.ts` which gets updated pre build with the `version` from `package.json`.

File: package.json (hunk 2 above)
Comment (dilico) [reply package.json]: Did you mean something like this?
`"start": "node dist/index.js"`

File: src/routing/index.ts (hunk 3 above)
Comment (dilico) [reply src/routing/index.ts]: Ok!

File: tsconfig.json (hunk 4 above)
Comment (dilico) [reply tsconfig.json:32]: See above.

File: src/index.ts (hunk 5 above)
Comment (dilico) [reply src/index.ts]: Ok!

File: src/routing/schedule-handler.ts (hunk 7 above)
Comment (dilico) [reply src/routing/schedule-handler.ts:15]: Yeah, I forgot it. Noticed it when I was doing the integration tests.

File: src/routing/schedule-handler.ts (hunk 8 above)
Comment (dilico) [reply src/routing/schedule-handler.ts:19]: Yep

File: src/services/fixtures/fixture.ts (hunk 9 above)
Comment (dilico) [reply src/services/fixtures/fixture.ts]: Ooops...

File: src/scheduler/index.ts (hunk 10 above)
Comment (dilico) [reply src/scheduler/index.ts]: Good point!

File: src/scheduler/index.ts (hunk 12 above)
Comment (dilico) [reply src/scheduler/index.ts]: Ok!

File: src/scheduler/index.ts (hunk 11 above)
Comment (dilico) [reply src/scheduler/index.ts]: Absolutely!

File: src/routing/pre-match-handler.ts (hunk 6 above)
Comment (dilico) [reply src/routing/pre-match-handler.ts]: I used promises to avoid wrapping every request handler in a `try catch` block. Any error that occurs will be picked up by the `catch` part of the promise call. Or am I missing something?
